File: src/ai/schemas.ts
import { z } from 'zod'

// === Character Schemas ===

export const PersonalitySchema = z.object({
  keywords: z.array(z.string()).default([]),
  surface: z.string().default(''),
  inner: z.string().default(''),
  stressResponse: z.string().default(''),
})

export const CreateCharacterSchema = z.object({
  name: z.string().min(1, '角色名称不能为空'),
  aliases: z.array(z.string()).default([]),
  tags: z.array(z.string()).default([]),
  appearance: z.string().default(''),
  personality: PersonalitySchema.default({}),
  background: z.string().default(''),
  trauma: z.string().optional(),
  goals: z.string().optional(),
  arc: z.string().optional(),
  quotes: z.array(z.string()).default([]),
  abilities: z.array(z.string()).default([]),
})

export const UpdateCharacterSchema = z.object({
  characterId: z.string().min(1, '角色ID不能为空'),
  changes: z.record(z.unknown()).refine((v) => Object.keys(v).length > 0, '变更内容不能为空'),
})

// === Plot Schemas ===

export const CreatePlotNodeSchema = z.object({
  title: z.string().min(1, '节点标题不能为空'),
  summary: z.string().default(''),
  content: z.string().optional(),
  type: z.enum(['trunk', 'branch', 'if', 'foreshadowing']).default('branch'),
  characters: z.array(z.string()).default([]),
  location: z.string().optional(),
  tags: z.array(z.string()).default([]),
  parentIds: z.array(z.string()).default([]),
  condition: z.string().optional(),
})

// === Relation Schemas ===

export const CreateRelationSchema = z.object({
  sourceId: z.string().min(1, '源角色ID不能为空'),
  targetId: z.string().min(1, '目标角色ID不能为空'),
  type: z.string().min(1, '关系类型不能为空'),
  description: z.string().default(''),
  isHidden: z.boolean().default(false),
})

// === System Schemas ===

export const SystemLevelSchema = z.object({
  name: z.string().min(1),
  description: z.string().default(''),
  abilities: z.array(z.string()).default([]),
  restrictions: z.array(z.string()).default([]),
})

export const SystemBranchSchema = z.object({
  name: z.string().min(1),
  levels: z.array(SystemLevelSchema).default([]),
})

export const SystemRuleSchema = z.object({
  description: z.string().min(1),
  severity: z.enum(['hard', 'soft']).default('soft'),
  exceptions: z.array(z.string()).default([]),
})

export const CreateSystemSchema = z.object({
  name: z.string().min(1, '体系名称不能为空'),
  description: z.string().default(''),
  branches: z.array(SystemBranchSchema).default([]),
  rules: z.array(SystemRuleSchema).default([]),
})

// === Idea Schemas ===

export const CreateIdeaSchema = z.object({
  content: z.string().min(1, '灵感内容不能为空'),
  tags: z.array(z.string()).default([]),
})

// === Tool Input Router ===

export function validateToolInput(
  toolName: string,
  args: Record<string, unknown>,
): { success: true; data: Record<string, unknown> } | { success: false; error: string } {
  try {
    let schema: z.ZodTypeType<Record<string, unknown>>

    switch (toolName) {
      case 'createCharacter':
        schema = CreateCharacterSchema
        break
      case 'updateCharacter':
        schema = UpdateCharacterSchema
        break
      case 'createPlotNode':
        schema = CreatePlotNodeSchema
        break
      case 'createRelation':
        schema = CreateRelationSchema
        break
      case 'createSystem':
        schema = CreateSystemSchema
        break
      case 'createIdea':
        schema = CreateIdeaSchema
        break
      default:
        return { success: true, data: args }
    }

    const result = schema.safeParse(args)
    if (!result.success) {
      const issues = result.error.issues.map((i) => `${i.path.join('.')}: ${i.message}`).join('; ')
      return { success: false, error: `参数校验失败: ${issues}` }
    }

    return { success: true, data: result.data as Record<string, unknown> }
  } catch (err) {
    return { success: false, error: `校验异常: ${err instanceof Error ? err.message : '未知错误'}` }
  }
}

File: backend/schemas.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any


class UserBase(BaseModel):
    username: str
    email: str


class UserCreate(UserBase):
    password: str = Field(..., min_length=6)


class UserOut(UserBase):
    id: int

    class Config:
        from_attributes = True


class Token(BaseModel):
    access_token: str
    token_type: str = "bearer"


class LoginRequest(BaseModel):
    username: str
    password: str


class WorkBase(BaseModel):
    id: str
    name: str
    genre: Optional[str] = None
    description: Optional[str] = None
    total_chapters: Optional[int] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class WorkCreate(BaseModel):
    id: str
    name: str
    genre: Optional[str] = None
    description: Optional[str] = None
    total_chapters: Optional[int] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class WorkOut(WorkBase):
    class Config:
        from_attributes = True


class Personality(BaseModel):
    keywords: List[str] = []
    surface: Optional[str] = None
    inner: Optional[str] = None
    stress_response: Optional[str] = None


class CharacterBase(BaseModel):
    id: str
    work_id: str
    name: str
    aliases: List[str] = []
    tags: List[str] = []
    appearance: Optional[str] = None
    personality: Optional[Personality] = None
    background: Optional[str] = None
    trauma: Optional[str] = None
    goals: Optional[str] = None
    arc: Optional[str] = None
    quotes: List[str] = []
    abilities: List[str] = []
    first_appearance: Optional[str] = None
    status: str = "alive"
    images: List[str] = []
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class CharacterCreate(CharacterBase):
    pass


class CharacterOut(CharacterBase):
    class Config:
        from_attributes = True


class Foreshadowing(BaseModel):
    id: str
    description: str
    status: str = "unresolved"


class Payoff(BaseModel):
    id: str
    description: str
    linked_foreshadowing: Optional[str] = None


class PlotNodeBase(BaseModel):
    id: str
    work_id: str
    title: str
    summary: Optional[str] = None
    content: Optional[str] = None
    type: str = "trunk"
    status: str = "todo"
    characters: List[str] = []
    location: Optional[str] = None
    tags: List[str] = []
    parent_ids: List[str] = []
    child_ids: List[str] = []
    condition: Optional[str] = None
    foreshadowing: List[Foreshadowing] = []
    payoff: List[Payoff] = []
    word_count_target: Optional[int] = None
    word_count_actual: Optional[int] = None
    position: Optional[Dict[str, Any]] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class PlotNodeCreate(PlotNodeBase):
    pass


class PlotNodeOut(PlotNodeBase):
    class Config:
        from_attributes = True


class RelationEdgeBase(BaseModel):
    id: str
    work_id: str
    source_id: str
    target_id: str
    type: str
    description: Optional[str] = None
    is_hidden: bool = False
    created_at: Optional[str] = None


class RelationEdgeCreate(RelationEdgeBase):
    pass


class RelationEdgeOut(RelationEdgeBase):
    class Config:
        from_attributes = True


class SystemBranch(BaseModel):
    id: str
    name: str
    levels: List[Dict[str, Any]] = []


class SystemRule(BaseModel):
    id: str
    description: str
    severity: str = "soft"
    exceptions: List[str] = []


class WorkSystemBase(BaseModel):
    id: str
    work_id: str
    name: str
    description: Optional[str] = None
    branches: List[SystemBranch] = []
    rules: List[SystemRule] = []
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class WorkSystemCreate(WorkSystemBase):
    pass


class WorkSystemOut(WorkSystemBase):
    class Config:
        from_attributes = True


class IdeaBase(BaseModel):
    id: str
    work_id: str
    content: str
    tags: List[str] = []
    status: str = "pending"
    linked_entity: Optional[Dict[str, Any]] = None
    created_at: Optional[str] = None


class IdeaCreate(IdeaBase):
    pass


class IdeaOut(IdeaBase):
    class Config:
        from_attributes = True


class StoryEventBase(BaseModel):
    id: str
    work_id: str
    title: str
    description: Optional[str] = None
    type: str = "other"
    chapter_id: Optional[str] = None
    character_ids: List[str] = []
    timestamp: Optional[str] = None
    location: Optional[str] = None
    importance: int = 1
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class StoryEventCreate(StoryEventBase):
    pass


class StoryEventOut(StoryEventBase):
    class Config:
        from_attributes = True


class EventEdgeBase(BaseModel):
    id: str
    work_id: str
    source_id: str
    target_id: str
    type: str = "sequence"
    description: Optional[str] = None
    created_at: Optional[str] = None


class EventEdgeCreate(EventEdgeBase):
    pass


class EventEdgeOut(EventEdgeBase):
    class Config:
        from_attributes = True
